File: 0621-task-scheduler/0621-task-scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        priority_queue<pair<int, char>> maxHeap;
        unordered_map<char, int> map;
        vector<pair<int, char>> temp;
        for(auto el: tasks){
            map[el]++;
        }
        for(auto el: map){
            maxHeap.push({el.second, el.first});
            map[el.first] = 0;
        }
        int counter = 0;
        while(!maxHeap.empty()){
            counter++;
            temp.clear();
            while(!maxHeap.empty() && map[maxHeap.top().second] >= counter){
                temp.push_back(maxHeap.top());
                maxHeap.pop();
            }
            if(!maxHeap.empty()){
                auto currentTask = maxHeap.top();
                maxHeap.pop();
                char taskId = currentTask.second;
                if(counter > map[taskId]){
                    currentTask.first--;
                    map[taskId] = counter + n;
                }
                if(currentTask.first){
                    maxHeap.push(currentTask);
                }
            }
            for(auto el: temp){
                maxHeap.push(el);
            }
        }
        return counter;
    }
};

File: 0621-task-scheduler/0621-task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        # freq_map = Counter(tasks)
        # max_heap = [-freq for freq in freq_map.values()]

        # heapq.heapify(max_heap)
        
        # q = deque()
        # counter = 0

        # while max_heap or q:
        #     counter += 1

        #     if max_heap:
        #         current_task = heappop(max_heap)
            
        #         # used current task
        #         current_task += 1

        #         if current_task < 0:
        #             q.append([counter + n, current_task])
            
        #     if q and counter == q[0][0]:
        #         heappush(max_heap, q.popleft()[1])

        # return counter

        if n == 0:
            return len(tasks)

        freq_map = Counter(tasks)
        freqs =  list(freq_map.values())
        freqs.sort()
        max_freq = freqs[-1]
        num_of_blanks = max_freq - 1
        num_of_idle_spots = num_of_blanks * n

        for freq in freqs[:-1]:
            num_of_idle_spots -= min(num_of_blanks, freq)
        

        if num_of_idle_spots > 0:
            return len(tasks) + num_of_idle_spots
        
        return len(tasks)
